projects/rocprofiler-sdk/source/lib/python/rocpd/source/perfetto.cpp

```diff
@@ -517,6 +517,102 @@ write_perfetto(
+            {
+                struct kernel_dispatch_group_index
+                {
+                    uint64_t agent_abs_index;
+                    uint64_t stream_id = 0;
+                    uint64_t queue_id  = 0;
+
+                    bool operator<(const kernel_dispatch_group_index& other) const
+                    {
+                        if(agent_abs_index != other.agent_abs_index)
+                            return agent_abs_index < other.agent_abs_index;
+
+                        if(queue_id != other.queue_id) return queue_id < other.queue_id;
+
+                        return stream_id < other.stream_id;
+                    }
+                };
+
+                struct kernel_dispatch_data
+                {
+                    std::vector<types::kernel_dispatch>::iterator sample;
+                    rocprofiler_timestamp_t                       timestamp;
+                };
+
+                std::map<kernel_dispatch_group_index, std::vector<kernel_dispatch_data>>
+                    kernel_groups;
+
+                for(auto it = begin(gen); it != end(gen); ++it)
+                {
+                    kernel_dispatch_group_index group_index;
+                    group_index.agent_abs_index = it->agent_abs_index;
+                    if(ocfg.group_by_queue)
+                        group_index.queue_id = it->queue_id;
+                    else
+                        group_index.stream_id = it->stream_id;
+
+                    kernel_groups[group_index].push_back({it, it->start / 2 + it->end / 2});
+                }
+
+                for(auto group_it = begin(kernel_groups); group_it != end(kernel_groups);
+                    ++group_it)
+                {
+                    auto& group_data = group_it->second;
+                    std::sort(begin(group_data),
+                              end(group_data),
+                              [&](const kernel_dispatch_data& a, const kernel_dispatch_data& b) {
+                                  return a.timestamp < b.timestamp;
+                              });
+
+                    for(auto sample_it = group_data.begin(); sample_it != group_data.end() - 1;
+                        ++sample_it)
+                    {
+                        auto next_sample_it = std::next(sample_it);
+
+                        auto current_it = sample_it->sample;
+                        auto next_it    = next_sample_it->sample;
+
+                        if(next_it->start >= current_it->end) continue;
+
+                        auto current_time = sample_it->timestamp;
+                        auto next_time    = next_sample_it->timestamp;
+
+                        auto   current_half_span = current_it->end - current_time;
+                        auto   next_half_span    = next_time - next_it->start;
+                        auto   total_span        = current_half_span + next_half_span;
+                        auto   max_span          = next_time - current_time;
```

Comment:
```suggestion
                        auto   current_mid_to_end_duration = current_it->end - current_time;
                        auto   next_start_to_mid_duration    = next_time - next_it->start;
                        auto   total_span        = current_half_span + next_half_span; // Total duration to redistribute
                        auto   max_span          = next_time - current_time; // Available space between midpoints
```
current_mid_to_end_duration  or current_mid_to_end_time?  A more descriptive name might help.  

Also, add some comments on what these calculations represent

Reply:
Done